### ISSUE TITLE
Lot size rounding orders to 0, then submitting 0-quantity orders.

### ISSUE THREAD

**Issue by jaredbroad** [QuantConnect/Lean#457]
Issue when submitting an order with quantity less than the lot size; then its rounded to 0 after the sanity checks. Then the 0-quantity order is submitted.

E.g. FXCM lot size 1k, submit order for 200 EUR, rounded to 0 but submitted anyway (and changed fees)

Should probably perform a non-zero check after rounding the order; &/or; round the lot size earlier in the process and check it still passes the non-zero check

![2016-07-06 18_09_58-algorithmic trading platform - quantconnect com](https://cloud.githubusercontent.com/assets/2243225/16639676/d3c934d4-43bf-11e6-8644-012dfd5591aa.png)

**Comment by mchandschuh** [comment]
Some comments/thoughts/recommendations:

The method `BrokerageTransactionHandler.RoundOffOrder` seems misplaced and hard coded. Recommend moving to concern to `IBrokerageModel` via a new method -- `IBrokerageModel.PrepareSubmitOrderRequest(Security, SubmitOrderRequest)` (naming suggestions welcomed) -- which would give the model a chance to modify the request as soon as its created in the algorithm. This will allow the behavior to be pluggable and then also be subject to the various sanity checks within the algorithm.

We don't call `algorithm.BrokerageModel.CanSubmitOrder` in live mode. The documentation for the interface method seems to indicate its there to support things such as connectivity and size restrictions. Recommend allowing this logic to execute even in live mode, which will provide better consistency between backtest/live modes. This does not directly impact this bug, but is worth noting nevertheless.

@jaredbroad Let me know what you think and I can get started on this.

**Comment by jaredbroad** [comment]
Agree RoundOffOrder was kind of randomly placed - but don't all orders pass through there? It was there since all brokerages needed rounding and rather than replicating it on each model it was only coded once (if done via interface-method on that IBrokerageModel that is fine too).

Agree re: CanSubmitOrder; certainly helps understand the error messages. Some of the brokerage errors are misleading so best if we filter them first. 

How would the "order rounded" messaging get back to the user?

**Comment by mchandschuh** [comment]
###### RE: Duplicate rounding logic
I believe all of our models derive from a `DefaultBrokerageModel` which would likewise define the default rounding behavior. So no duplication really needed.
###### RE: Order rounded messasge
Great point. I hadn't previously considered this. We could modify the proposed signature to `PrepareSubmitOrderRequest(Security, SubmitOrderRequest, out BrokerageMessageEvent)`. This is similar to the pattern used on `CanSubmitOrder(Security, Order, out BrokerageMessageEvent)`.

**Comment by mchandschuh** [comment]
This issue is fixed via https://github.com/QuantConnect/Lean/commit/b839705a99377b4a6aef81321cb567eab2f58b2c... do we still want to make any changes here?

**Comment by jaredbroad** [comment]
This fix has a bug somewhere -- the order status isn't getting set properly and so there's no feedback the order was rejected + the SecurityManager.WaitForOrder() is timing out since the order never returns invalid.  
```
namespace QuantConnect 
{   
    public class BasicTemplateAlgorithm : QCAlgorithm
    {
    	int attempts = 0;
        public override void Initialize() 
        {
            SetStartDate(2015, 6, 1);         
            SetEndDate(2016,1,1);
            SetCash(25000);
            AddForex("EURUSD", Resolution.Minute);
        } 
        public void OnData(TradeBars data) 
        {   
            if (!Portfolio.HoldStock && attempts++ < 3) 
            {
                Order("EURUSD",  200);
            }
        }
    }
}
```

**Comment by Phoenix1271** [comment]
Timeout issue fixed in #483

**Comment by jameschch** [comment]
I noticed the problem with zero quantities slipping though when working on another issue. My interim solution was to change QCAlgorithm.PreOrderChecksImpl to have an additional clause:

```
if (request.Quantity == 0 || request.Symbol == null || request.Symbol == QuantConnect.Symbol.Empty || Math.Abs(request.Quantity) < Securities[request.Symbol].SymbolProperties.LotSize)
{
                return OrderResponse.ZeroQuantity(request);
}
```
I think this is run in backtest and live and already has a means to provides a warning. Is this of any help for this issue?

**Comment by jaredbroad** [comment]
"Math.Abs(request.Quantity) < Securities[request.Symbol].
SymbolProperties.LotSize"

That is a good idea :) Much simpler.

On Wed, Jul 13, 2016 at 5:12 AM, jameschch <notifications@github.com> wrote:

> I noticed the problem with zero quantities slipping though when working on
> another issue. My interim solution was to change
> QCAlgorithm.PreOrderChecksImpl to have an additional clause:
>
> if (request.Quantity == 0 || request.Symbol == null || request.Symbol == QuantConnect.Symbol.Empty || Math.Abs(request.Quantity) < Securities[request.Symbol].SymbolProperties.LotSize)
> {
>                 return OrderResponse.ZeroQuantity(request);
> }
>
> I think this is run in backtest and live and already has a means to
> provides a warning. Is this of any help for this issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/QuantConnect/Lean/issues/457#issuecomment-232300901>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACI6mUn0pzYk0Hq205TsG5grNkg3Lzypks5qVKv2gaJpZM4JGp2q>
> .
>



-- 
Jared Broad

**Comment by jaredbroad** [comment]
Fixed in various commits 003e490, 1642aed, b4402a4
